Firestore/core/src/firebase/firestore/nanopb/message.h

```diff
@@ -198,6 +195,23 @@ Message<T> Message<T>::TryParse(Reader* reader) {
   return result;
 }
 
+/**
+ * Deduces the type of the `Message` from the given initializer, similarly to
+ * `std::make_pair`.
+ */
+// The resulting message takes ownership of the given proto, so only rvalue
+// references should be accepted. Unfortunately, because this function is
+// a template, the argument type is actually a forwarding reference, which would
+// bind to lvalues and rvalues alike. To work around this, use some
+// metaprogramming.
+template <typename T,
+          typename = absl::enable_if_t<!std::is_lvalue_reference<T>::value>>
+Message<T> MakeMessage(T&& proto) {
```

Comment:
Given this method has exactly one caller and the complexity of the metaprogramming required to implement it, is this really worth it?

I guess I can see that this might be commonly used in combination with the serializer, though I think it leads to some potentially hard to reason about code:

```c++
auto message = MakeMessage(serializer_.EncodeDocument(...));
```

Should we add a move-assignment `operator=` so that we could write this?

```c++
Message<google_firestore_v1_Document> message = serializer_.EncodeDocument(...);
```